File: examples/siglip2/python/siglip2.py
#!/usr/bin/env python3
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
# SPDX-License-Identifier: Apache-2.0

import argparse
from pathlib import Path

import cv2
import yaml

from spacemit_vision import VisionServiceNative, VisionServiceStatus


def resolve_path(path_value, project_root):
    p = Path(path_value).expanduser()
    return p if p.is_absolute() else (project_root / p).resolve()


def load_labels_file(path):
    labels = []
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            labels.append(line)
    return labels


def parse_args():
    parser = argparse.ArgumentParser(description="SigLIP2 image-text embedding example")
    parser.add_argument("--config", type=str, default=None)
    parser.add_argument("--image", type=str, default=None)
    parser.add_argument(
        "--labels",
        type=str,
        default=None,
        help="Scene labels file (default: scene_labels_path from config)",
    )
    parser.add_argument(
        "--text",
        type=str,
        default=None,
        help="Comma-separated text prompts (overrides --labels)",
    )
    parser.add_argument("--topk", type=int, default=5)
    parser.add_argument("--model-path", type=str, default=None)
    return parser.parse_args()


def main():
    args = parse_args()
    default_config = Path(__file__).parent.parent / "config" / "siglip2.yaml"
    config_path = Path(args.config) if args.config else default_config
    project_root = Path(__file__).parent.parent.parent.parent

    config = {}
    if config_path.exists():
        with open(config_path, "r", encoding="utf-8") as f:
            config = yaml.safe_load(f) or {}

    image_path = args.image or config.get("test_image")
    if not image_path:
        raise ValueError("Provide --image or set test_image in config")
    image_path = resolve_path(image_path, project_root)

    if args.text:
        labels = [t.strip() for t in args.text.split(",") if t.strip()]
    else:
        labels_path = args.labels or config.get("scene_labels_path")
        if labels_path:
            labels = load_labels_file(resolve_path(labels_path, project_root))
        else:
            labels = ["a photo of a dog", "a photo of a cat", "a photo of a car"]

    model_path_override = ""
    if args.model_path:
        p = Path(args.model_path).expanduser()
        model_path_override = str(p if p.is_absolute() else (project_root / p).resolve())

    svc = VisionServiceNative.create(str(config_path), model_path_override=model_path_override)

    image = cv2.imread(str(image_path))
    if image is None:
        raise FileNotFoundError(f"Failed to load image: {image_path}")

    st, image_emb = svc.infer_embedding(image)
    if st != VisionServiceStatus.OK:
        raise RuntimeError(svc.last_error())

    scored = []
    print(f"Image: {image_path}")
    for label in labels:
        st, text_emb = svc.encode_text(label)
        if st != VisionServiceStatus.OK:
            raise RuntimeError(svc.last_error())
        score = VisionServiceNative.embedding_similarity(image_emb, text_emb)
        scored.append((label, score))

    scored.sort(key=lambda x: x[1], reverse=True)
    print(f"Top {min(args.topk, len(scored))} matches:")
    for label, score in scored[: args.topk]:
        print(f"  {label} : {score:.4f}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/siglip2/cpp/siglip2.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <algorithm>
#include <fstream>
#include <iostream>  // NOLINT(build/include_order)
#include <iomanip>   // NOLINT(build/include_order)
#include <memory>    // NOLINT(build/include_order)
#include <sstream>   // NOLINT(build/include_order)
#include <string>    // NOLINT(build/include_order)
#include <variant>   // NOLINT(build/include_order)
#include <vector>    // NOLINT(build/include_order)

#include "vision_service.h"  // NOLINT(build/include_order)

static std::vector<std::string> split_csv(const std::string& csv) {
    std::vector<std::string> out;
    std::stringstream ss(csv);
    std::string item;
    while (std::getline(ss, item, ',')) {
        const auto start = item.find_first_not_of(" \t");
        const auto end = item.find_last_not_of(" \t");
        if (start == std::string::npos) {
            continue;
        }
        out.push_back(item.substr(start, end - start + 1));
    }
    return out;
}

static std::vector<std::string> load_labels_file(const std::string& path) {
    std::ifstream file(path);
    if (!file.is_open()) {
        throw std::runtime_error("Failed to open labels file: " + path);
    }
    std::vector<std::string> labels;
    std::string line;
    while (std::getline(file, line)) {
        while (!line.empty() && (line.back() == '\r' || line.back() == ' ' || line.back() == '\t')) {
            line.pop_back();
        }
        if (line.empty() || line[0] == '#') {
            continue;
        }
        labels.push_back(line);
    }
    return labels;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: " << argv[0] << " <config_yaml> [options]\n"
            << "  Options: --image <path> --labels <scene_labels.txt>\n"
            << "           --text \"label1,label2\" --topk N --model-path <path>"
            << std::endl;
        return 1;
    }

    std::string config_path = argv[1];
    std::string image_path;
    std::string labels_path;
    std::string text_csv;
    int topk = 5;
    std::string model_path_override;

    for (int i = 2; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--image" && i + 1 < argc) {
            image_path = argv[++i];
        } else if (arg == "--labels" && i + 1 < argc) {
            labels_path = argv[++i];
        } else if (arg == "--text" && i + 1 < argc) {
            text_csv = argv[++i];
        } else if (arg == "--topk" && i + 1 < argc) {
            topk = std::stoi(argv[++i]);
        } else if (arg == "--model-path" && i + 1 < argc) {
            model_path_override = argv[++i];
        }
    }

    std::unique_ptr<VisionService> service =
        VisionService::Create(config_path, model_path_override, true);
    if (!service) {
        std::cerr << "Error: " << VisionService::LastCreateError() << std::endl;
        return 1;
    }

    if (image_path.empty()) {
        image_path = service->GetDefaultImage();
    }
    if (image_path.empty()) {
        std::cerr << "Error: No image path. Use --image or set test_image in config." << std::endl;
        return 1;
    }

    std::vector<std::string> labels;
    if (!text_csv.empty()) {
        labels = split_csv(text_csv);
    } else {
        if (labels_path.empty()) {
            labels_path = service->GetConfigPathValue("scene_labels_path");
        }
        if (!labels_path.empty()) {
            try {
                labels = load_labels_file(labels_path);
            } catch (const std::exception& e) {
                std::cerr << "Error: " << e.what() << std::endl;
                return 1;
            }
        }
    }
    if (labels.empty()) {
        labels = {"a photo of a dog", "a photo of a cat", "a photo of a car"};
    }

    VisionServiceResponse response;
    if (service->Infer(image_path, &response) != VISION_SERVICE_OK) {
        std::cerr << "Error: " << service->LastError() << std::endl;
        return 1;
    }
    if (response.results.empty()) {
        std::cerr << "Error: No image embedding returned." << std::endl;
        return 1;
    }
    const vision::Embedding* image_emb = std::get_if<vision::Embedding>(&response.results[0]);
    if (image_emb == nullptr) {
        std::cerr << "Error: Model did not return an embedding." << std::endl;
        return 1;
    }

    std::vector<std::pair<std::string, float>> scored;
    scored.reserve(labels.size());
    std::cout << "Image: " << image_path << std::endl;
    for (const std::string& label : labels) {
        std::vector<float> text_emb;
        if (service->EncodeText(label, &text_emb) != VISION_SERVICE_OK) {
            std::cerr << "Error encoding text: " << service->LastError() << std::endl;
            return 1;
        }
        const float score = VisionService::EmbeddingSimilarity(image_emb->embedding, text_emb);
        scored.emplace_back(label, score);
    }

    std::sort(scored.begin(), scored.end(),
        [](const auto& a, const auto& b) { return a.second > b.second; });
    const int show_k = std::min(topk, static_cast<int>(scored.size()));
    std::cout << "Top " << show_k << " matches:" << std::endl;
    for (int i = 0; i < show_k; ++i) {
        std::cout << "  " << scored[static_cast<size_t>(i)].first << " : "
            << std::fixed << std::setprecision(4) << scored[static_cast<size_t>(i)].second
            << std::endl;
    }
    return 0;
}
